File: apps/mobile/test/protocol_version_test.dart
import 'package:ccpocket/models/protocol_version.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('ProtocolCompatibility', () {
    test('treats a Bridge without metadata as protocol 1', () {
      final compatibility = ProtocolCompatibility.forBridge();

      expect(compatibility.isCompatible, isTrue);
      expect(compatibility.selectedVersion, 1);
      expect(compatibility.assumedLegacyBridge, isTrue);
    });

    test('selects the highest overlapping protocol version', () {
      final compatibility = ProtocolCompatibility.forBridge(
        minimumProtocolVersion: 1,
        protocolVersion: 1,
      );

      expect(compatibility.isCompatible, isTrue);
      expect(compatibility.selectedVersion, 1);
      expect(compatibility.assumedLegacyBridge, isFalse);
    });

    test('rejects a non-overlapping Bridge protocol range', () {
      final compatibility = ProtocolCompatibility.forBridge(
        minimumProtocolVersion: 2,
        protocolVersion: 2,
      );

      expect(compatibility.isCompatible, isFalse);
      expect(compatibility.selectedVersion, isNull);
      expect(compatibility.updateTarget, ProtocolUpdateTarget.app);
    });

    test('rejects malformed Bridge protocol metadata', () {
      final compatibility = ProtocolCompatibility.forBridge(
        minimumProtocolVersion: 1,
      );

      expect(compatibility.isCompatible, isFalse);
      expect(compatibility.malformedBridgeRange, isTrue);
      expect(compatibility.updateTarget, ProtocolUpdateTarget.both);
    });

    test('rejects malformed Bridge JSON declarations fail-closed', () {
      for (final declaration in <Map<String, dynamic>>[
        {'minimumProtocolVersion': 1},
        {'protocolVersion': 0},
        {'protocolVersion': -1},
        {'protocolVersion': 1.0},
        {'protocolVersion': '1'},
        {'minimumProtocolVersion': 2, 'protocolVersion': 1},
      ]) {
        final compatibility = ProtocolCompatibility.fromBridgeJson(declaration);

        expect(
          compatibility.isCompatible,
          isFalse,
          reason: '$declaration must not bypass protocol negotiation',
        );
      }
    });

    test('treats malformed rejection ranges as an update-both failure', () {
      for (final declaration in <Map<String, dynamic>>[
        <String, dynamic>{},
        {'minimumProtocolVersion': 1},
        {'protocolVersion': '2', 'minimumProtocolVersion': 1},
        {'protocolVersion': 1, 'minimumProtocolVersion': 2},
        {'protocolVersion': 0, 'minimumProtocolVersion': 0},
      ]) {
        final compatibility = ProtocolCompatibility.fromBridgeRejectionJson(
          declaration,
        );

        expect(compatibility.isCompatible, isFalse);
        expect(compatibility.malformedBridgeRange, isTrue);
        expect(compatibility.updateTarget, ProtocolUpdateTarget.both);
      }
    });
  });
}

File: packages/bridge/src/protocol-version.test.ts
import { describe, expect, it } from "vitest";
import {
  clientProtocolRange,
  negotiateProtocolVersion,
} from "./protocol-version.js";

describe("protocol version negotiation", () => {
  it("treats clients without metadata as protocol 1", () => {
    expect(clientProtocolRange({})).toEqual({ min: 1, max: 1 });
  });

  it("treats a singular client version as an exact range", () => {
    expect(clientProtocolRange({ protocolVersion: 2 })).toEqual({
      min: 2,
      max: 2,
    });
  });

  it("selects the highest overlapping protocol version", () => {
    expect(
      negotiateProtocolVersion({ min: 1, max: 3 }, { min: 2, max: 4 }),
    ).toBe(3);
  });

  it("returns null when protocol ranges do not overlap", () => {
    expect(
      negotiateProtocolVersion({ min: 1, max: 1 }, { min: 2, max: 2 }),
    ).toBeNull();
  });
});
